rpm/0011-Add-profile-files-to-a-list-when-processing-argument.patch

```diff
@@ -0,0 +1,149 @@
+
+Do not read the profile files when processing arguments. Instead append
+them to a list that is processed after the arguments are processed in
+order to be able to replace all template keys in the profile file lines.
+Each item in the list is free'd after it has been processed.
+---
+ src/firejail/firejail.h |  1 +
+ src/firejail/main.c     |  3 ++
+ src/firejail/profile.c  | 73 +++++++++++++++++++++++++++++++++++++++++
+ 3 files changed, 77 insertions(+)
+
+diff --git a/src/firejail/firejail.h b/src/firejail/firejail.h
+index e4eebe62..5ac874e0 100644
+--- a/src/firejail/firejail.h
++++ b/src/firejail/firejail.h
+@@ -454,6 +454,7 @@ char *profile_list_slice(char *pos, char **ppos);
+ char *profile_list_normalize(char *list);
+ char *profile_list_compress(char *list);
+ void profile_list_augment(char **list, const char *items);
++void profile_read_file_list();
+ 
+ // list.c
+ void list(void);
+diff --git a/src/firejail/main.c b/src/firejail/main.c
+index 4a367d7e..88b99de7 100644
+--- a/src/firejail/main.c
++++ b/src/firejail/main.c
+@@ -2732,6 +2732,9 @@ int main(int argc, char **argv, char **envp) {
+ 			break;
+ 		}
+ 	}
++
++	profile_read_file_list();
++
+ 	EUID_ASSERT();
+ 
+ 	// exit chroot, overlay and appimage sandboxes when caps are explicitly specified on command line
+diff --git a/src/firejail/profile.c b/src/firejail/profile.c
+index c57d8d26..d4bbef97 100644
+--- a/src/firejail/profile.c
++++ b/src/firejail/profile.c
+@@ -26,6 +26,18 @@ extern char *xephyr_screen;
+ 
+ #define MAX_READ 8192				  // line buffer for profile files
+ 
++typedef struct profile_file_name_t {
++	char *fname;
++	struct profile_file_name_t *next;
++} ProfileFileName;
++
++// This is initially set to make profile_read() to add the profile filename
++// to a list of profiles that are to be read after arguments have been
++// processed and templates are set in order to replace any template key
++// existing in the profile lines.
++static int read_profiles = 0;
++static ProfileFileName *profile_file_name_list = NULL;
++
+ // find and read the profile specified by name from dir directory
+ // return  1 if a profile was found
+ static int profile_find(const char *name, const char *dir, int add_ext) {
+@@ -1643,6 +1655,38 @@ void profile_add(char *str) {
+ 	ptr->next = prf;
+ }
+ 
++// Appends entries to profile_file_name_list for later reading of the files
++void add_to_profile_file_name_list(const char *fname)
++{
++	ProfileFileName *pfn;
++	ProfileFileName *iter;
++
++	if (!fname || !*fname)
++		return;
++
++	if (arg_debug)
++		printf("add profile \"%s\" to list\n", fname);
++
++	pfn = malloc(sizeof(ProfileFileName));
++	if (!pfn)
++		errExit("malloc");
++
++	pfn->fname = strdup(fname);
++	pfn->next = NULL;
++
++	if (!profile_file_name_list) {
++		profile_file_name_list = pfn;
++		return;
++	}
++
++	// Add to the end as order of the profiles may be relevant.
++	iter = profile_file_name_list;
++	while (iter->next)
```

Comment:
This O(N^2) time complexity makes me wonder if you should either a) keep the end of the list somewhere or b) just insert to the beginning and then reverse the list (or alternatively iterate to the end and back while consuming it).

Reply:
Yep, that is similar approach as with the rest of the lists here in Firejail. I had in mind to change this but other changes apparently made me forget this. Might as well do that now to keep the last pointer there. Prepending caused the problem of nothing working, hence the order.

Reply:
Decided to use prepend items and then to reverse the list when reading the items from it to keep the current order. I guess this is cleaner way than to drag the additional pointer for the last item in the list.